File: common/src/main/resources/ts/models/QuestionSpecificFields.ts
import {Question} from "@common/models";

export class QuestionSpecificFields {
    id: number
    question_id: number;
    cursor_min_val: number;
    cursor_max_val: number;
    cursor_step: number;
    cursor_min_label: string;
    cursor_max_label: string;

    constructor(questionId?: number) {
        this.question_id = questionId;
        this.cursor_min_val = null;
        this.cursor_max_val = null;
        this.cursor_step = null;
        this.cursor_min_label = null;
        this.cursor_max_label = null;
    }

    build(data: IQuestionSpecificFieldsResponse) : QuestionSpecificFields {
        this.id = data.id ? data.id : null;
        this.question_id = data.questionId != null ? data.questionId : null;
        this.cursor_min_val = data.cursorMinVal != null ? data.cursorMinVal : null;
        this.cursor_max_val = data.cursorMaxVal != null ? data.cursorMaxVal : null;
        this.cursor_step = data.cursorStep != null ? data.cursorStep : null;
        this.cursor_min_label = data.cursorMinLabel ? data.cursorMinLabel : null;
        this.cursor_max_label = data.cursorMaxLabel ? data.cursorMaxLabel : null;
        return this;
    }

    toJson() : Object {
        return {
            id: this.id,
            question_id: this.question_id,
            cursor_min_val: this.cursor_min_val,
            cursor_max_val: this.cursor_max_val,
            cursor_step: this.cursor_step,
            cursor_min_label: this.cursor_min_label,
            cursor_max_label: this.cursor_max_label
        }
    }
}

export class QuestionSpecificFieldsPayload {
    id: number
    question_id: number;
    cursor_min_val: number;
    cursor_max_val: number;
    cursor_step: number;
    cursor_min_label: string;
    cursor_max_label: string;

    constructor(question: Question) {
        if (!question.specific_fields) return;
        this.cursor_min_val = question.specific_fields.cursor_min_val != null ? question.specific_fields.cursor_min_val : null;
        this.cursor_max_val = question.specific_fields.cursor_max_val != null ? question.specific_fields.cursor_max_val : null;
        this.cursor_step = question.specific_fields.cursor_step != null ? question.specific_fields.cursor_step : null;
        this.cursor_min_label = question.specific_fields.cursor_min_label != null ? question.specific_fields.cursor_min_label : null;
        this.cursor_max_label = question.specific_fields.cursor_max_label != null ? question.specific_fields.cursor_max_label : null;
    }

    toJson() : Object {
        return {
            id: this.id,
            question_id: this.question_id,
            cursor_min_val: this.cursor_min_val,
            cursor_max_val: this.cursor_max_val,
            cursor_step: this.cursor_step,
            cursor_min_label: this.cursor_min_label,
            cursor_max_label: this.cursor_max_label
        }
    }
}

export interface IQuestionSpecificFieldsResponse {
    id: number;
    questionId: number;
    cursorMinVal: number;
    cursorMaxVal: number;
    cursorStep: number;
    cursorMinLabel: string;
    cursorMaxLabel: string;
}

File: common/src/main/java/fr/openent/form/core/models/QuestionSpecificFields.java
package fr.openent.form.core.models;

import fr.openent.form.helpers.IModelHelper;
import io.vertx.core.json.JsonObject;

import java.util.Optional;

import static fr.openent.form.core.constants.Fields.*;

public class QuestionSpecificFields implements IModel<QuestionSpecificFields> {
    private Long id;
    private Long questionId;
    private Long cursorMinVal;
    private Long cursorMaxVal;
    private Long cursorStep;
    private String cursorMinLabel;
    private String cursorMaxLabel;


    // Constructors

    public QuestionSpecificFields() {
    }

    public QuestionSpecificFields(JsonObject questionSpecificFields) {
        this.id = questionSpecificFields.getLong(ID, null);
        this.questionId = questionSpecificFields.getLong(QUESTION_ID, null);
        this.cursorMinVal = questionSpecificFields.getLong(CURSOR_MIN_VAL,1L);
        this.cursorMaxVal = questionSpecificFields.getLong(CURSOR_MAX_VAL,10L);
        this.cursorStep = questionSpecificFields.getLong(CURSOR_STEP, 1L);
        this.cursorMinLabel = questionSpecificFields.getString(CURSOR_MIN_LABEL,"");
        this.cursorMaxLabel = questionSpecificFields.getString(CURSOR_MAX_LABEL,"");
    }


    // Getters

    public Long getId() { return id; }

    public Long getQuestionId() { return questionId; }

    public Long getCursorMinVal() { return cursorMinVal; }

    public Long getCursorMaxVal() { return cursorMaxVal; }

    public Long getCursorStep() { return cursorStep; }

    public String getCursorMinLabel() { return cursorMinLabel; }

    public String getCursorMaxLabel() { return cursorMaxLabel; }


    // Setters

    public QuestionSpecificFields setId(Long id) {
        this.id = id;
        return this;
    }

    public QuestionSpecificFields setQuestionId(Long questionId) {
        this.questionId = questionId;
        return this;
    }

    public QuestionSpecificFields setCursorMinVal(Long cursorMinVal) {
        this.cursorMinVal = cursorMinVal;
        return this;
    }

    public QuestionSpecificFields setCursorMaxVal(Long cursorMaxVal) {
        this.cursorMaxVal = cursorMaxVal;
        return this;
    }

    public QuestionSpecificFields setCursorStep(Long cursorStep) {
        this.cursorStep = cursorStep;
        return this;
    }

    public QuestionSpecificFields setCursorMinLabel(String cursorMinLabel) {
        this.cursorMinLabel = cursorMinLabel;
        return this;
    }

    public QuestionSpecificFields setCursorMaxLabel(String cursorMaxLabel) {
        this.cursorMaxLabel = cursorMaxLabel;
        return this;
    }


    // Functions

    public JsonObject toJson() {
        return IModelHelper.toJson(this, true, true);
    }

    /**
     * @deprecated Should instead use IModelHelper directly
     */
    @Deprecated
    @Override
    public QuestionSpecificFields model(JsonObject questionSpecificFields){
        return IModelHelper.toModel(questionSpecificFields, QuestionSpecificFields.class).orElse(null);
    }
}

